### ISSUE TITLE
Sellables endpoint error on ~200 pages in the API

### ISSUE THREAD

**Issue by Zebra447** [spreadshirt/shop-api-example-integration#1]
The sellables endpoint is broken. Almost all pages (17 to 208) are returning an error.

Examples: 
https://api.spreadshirt.com/api/v1/shops/266497/sellables?page=17&apiKey=XXXX
https://api.spreadshirt.com/api/v1/shops/266497/sellables?page=50&apiKey=XXXX
https://api.spreadshirt.com/api/v1/shops/266497/sellables?page=100&apiKey=XXXX
https://api.spreadshirt.com/api/v1/shops/266497/sellables?page=150&apiKey=XXXX
https://api.spreadshirt.com/api/v1/shops/266497/sellables?page=200&apiKey=XXXX
https://api.spreadshirt.com/api/v1/shops/266497/sellables?page=208&apiKey=XXXX

Error:
`{"message":"ProductType [460] could not be loaded"}`

**Comment by thimo1** [comment]
You can check free WordPress Plugin called WP-Spreadplugin it’s maintained and supports „Topics“ as well as everything else.
The Spreadshirt Forum is a good place asking the questions because there are many devs watching.

**Comment by thimo1** [comment]
For simply fetching I use the open shop.json which each shop is spitting out by default. Just open a spreadshop with enabled browser inspector. Then you can see the json resources it’s fetching.

**Comment by jsfi** [comment]
We strongly advise against using APIs that are not specified as public because they can change at any point in time without notice.

Thank you for bringing the paging issue to our attention. A fix for this problem will be released today.

**Comment by Zebra447** [comment]
Thanks for the suggestion, I didn't realize your plugin is not using the official API to fetch the sellables, but I just looked at your code and I see what you mean.

We will try getting the data from shop.json but it's a bit sketchy to use a data source that is not officially supported. It could break the entire website whenever Spreadshirt decides to change something without warning. Spreadshirt should just fix the API, but your solution is still useful as a temporary workaround - so thanks for this.

Edit: just saw jsfi's reply. I will wait for the fix. Thanks so much for the fast answer!

**Comment by thimo1** [comment]
> Spreadshirt should just fix the API, but your solution is still useful as a temporary workaround - so thanks for this.

Yeah I mean, it's always good to have a separate cache/database which you update regularly instead of accessing the data live. So changes to APIs won't hit you as long as it's not the basket API, which is, what's great, not changed so far.

**Comment by Zebra447** [comment]
Looks like the bug was fixed today - the API is not returning errors now

However, it seems like the total item count is wrong.

```
"count": 9504,
"limit": 48,
```

I suspect the counter wasn't updated after deleting the discontinued organic t-shirts which were causing the bug. I think the total count was the same number of 9504 yesterday.

When looping through the sellables, we get a total number of **8779** items and not 9504, so it creates issues with pagination when trying to get the number of pages by calculating 9504 divided by 48

Thanks for your help

**Comment by Zebra447** [comment]
> Looks like the bug was fixed today - the API is not returning errors now
> 
> However, it seems like the total item count is wrong.
> 
> ```
> "count": 9504,
> "limit": 48,
> ```
> 
> I suspect the counter wasn't updated after deleting the discontinued organic t-shirts which were causing the bug. I think the total count was the same number of 9504 yesterday.
> 
> When looping through the sellables, we get a total number of **8779** items and not 9504, so it creates issues with pagination when trying to get the number of pages by calculating 9504 divided by 48
> 
> Thanks for your help

The bug is still not fully fixed :(

**Comment by Zebra447** [comment]
After talking with Beatrice we just learned why we are facing this issue. The API has a limit of **maximum 9504 results** returned. I can't believe nobody shared this information until 2 weeks before the deadline. Why is this information not even available in the official documentation?

The "solution" suggested by Spreadshirt: take our 210,000 sellables and split them between 22 different shops to be below the 9504 limit ... and then somehow find a way to keep track of our designs through this mess. LOL!


I have some questions for @thimo1 - according to your experience, do you think your solution using the Spreadshop JSON "shopData" can bypass the limitations? Can it return more than 9504 sellables per shop?

If I understand correctly...
1. You get the list of articles from https://shop.spreadshirt.com/xxxxx/shopData/list
2. You get the Idea ID (sic) from the json, in the ID field
3. It shows the numberOfSellables but the sellable IDs are not returned on this page, how do you get it? 
4. There is not any officially supported "ideas" resource in the API but in your code you use https://api.spreadshirt.com/api/v1/ideas/IDEA_ID/sellables/SELLABLE_ID/  is there a way to get the list of Sellables ID corresponding to a Specific Idea ID?

At this point, we're open to using non-official workarounds because the official API has fundamental flaws and our shops will be a complete mess.

Thanks

**Comment by thimo1** [comment]
Hi,

I made the API adjustments to the plugin before the official API description was available. I did a lot of reverse engineering to put that up, and, it's still working.
After talking with some developers in January 2019, they told me, it isn't best practice to store sellables. That's the reason I switched to create them when a customer adds a product to the basket. That's the moment where a product configuration, as I call a sellable, is fixed. 
So, a color choice, size, product type and so on is stored as a sellable for later usage.

TL;DR 
That's why you see starting from line 2337, I'm creating a sellable on the fly while adding the product to the basket resource:
`
$apiUrlBase = 'https://api.spreadshirt.' . self::$shopOptions['shop_source'] . '/api/v1/ideas/' . self::replaceUnsecure($data['ideaId']) . '/sellables/' . self::replaceUnsecure($data['sellableId']) . '/product?appearanceId=' . (int) $data['appearance'] . '&mediaType=json&locale=' . (empty(self::$shopOptions['shop_language']) ? get_locale() : self::$shopOptions['shop_language']);
`

...with the appropriate payload.

3) To be honest, I never looked at the limit, but there's a pagination, so you may try yourself. I never use such a lot of products. I guess if there are so many products and caches are updated correctly @spreadshirt, the products should be displayed. But please keep in mind, some of the products you have in the old shop, might not be in the new shop, because they're end-of-life products or similar. If you can narrow down the missing products to a special product type, then that might be the reason (just a guess).
4) see above

Please be patient. I know the team of Spreadshirt is working hard to help you out and of course, if there are bugs, they will fix them in time.

Cheers
Thimo

**Comment by Zebra447** [comment]
Thanks for the details, it's helpful!

> I guess if there are so many products and caches are updated correctly @spreadshirt, the products should be displayed.


This is my point exactly. But I was told the backend will crash if it returns more than 9504 results. Spreadshirt needs to implement proper caching to fix the issue properly. Returning 200k results with proper caching & pagination shouldn't crash any server. Caching makes more sense than a sketchy workaround like splitting the designs among 22 shops. If there are still issues even with caching, they could add rate-limiting instead of a hard limit on the results.

> Please be patient. I know the team of Spreadshirt is working hard to help you out and of course, if there are bugs, they will fix them in time.

Maybe I sound annoying in the last days (I'm sorry) but in the last 10yrs, I was always appreciative of the support from the team. Initially, I was very patient (i started reporting the issues in August 2019). But with the deadline, we have only 2 weeks left and there's no way any human could manually sort 210,000 sellables among 22 shops in only 2 weeks. If our shop is broken due to bugs in the API, it's going to destroy 10 years of work and nobody seems to understand that :( It made no sense at all to put a deadline BEFORE fixing the issues. The new API should have been beta-tested before pushing it live...
